java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java

```
@@ -71,83 +75,126 @@
       throws Exception {
-    // Create a new instance of ClientIncomingAuthHeaderMiddleware.Factory. This factory creates
-    // new instances of ClientIncomingAuthHeaderMiddleware. The middleware processes
-    // username/password and bearer token authorization header authentication for this Flight Client.
-    final ClientIncomingAuthHeaderMiddleware.Factory factory =
-        new ClientIncomingAuthHeaderMiddleware.Factory(new ClientBearerHeaderHandler());
-
-    // Adds ClientIncomingAuthHeaderMiddleware.Factory instance to the FlightClient builder.
-    final FlightClient.Builder clientBuilder = FlightClient.builder();
-    if (verifyServer) {
-      clientBuilder
-          .allocator(allocator)
-          .location(Location.forGrpcTls(host, port))
-          .intercept(factory)
-          .useTls()
-          .verifyServer(false);
-    } else {
-      clientBuilder
-          .allocator(allocator)
-          .location(Location.forGrpcTls(host, port))
-          .intercept(factory)
-          .useTls()
-          .trustedCertificates(EncryptedConnectionUtils.getCertificateStream(
-              keyStorePath, keyStorePass));
-    }
 
-    final FlightClient client = clientBuilder.build();
-    return new AdhocFlightClient(client, allocator, authenticate(client, user, pass, factory, clientProperties));
+    return getClientHelper(
+      allocator,
+      host, port,
+      user, pass,
+      patOrAuthToken,
+      clientProperties,
+      () -> {
+        FlightClient.Builder flightClientBuilder = FlightClient.builder()
+            .location(Location.forGrpcTls(host, port))
+            .useTls();
+
+          if (verifyServer) {
+            flightClientBuilder.verifyServer(false);
+          } else {
+            try {
+              flightClientBuilder
+                  .trustedCertificates(EncryptedConnectionUtils.getCertificateStream(keyStorePath, keyStorePass));
+            } catch (Exception e) {
+              e.printStackTrace();
```

We shouldn't print the stack trace and swallow the exception here. If the problem is Supplier doesn't let you throw exceptions, just define a functional interface (one method interface annotated @FunctionalInterface) that lets you throw an exception.

---

Actually on second thought, it looks like you can just make getClientHelper take in a FlightClient.Builder instead of a Supplier to one.